GVFS/GVFS.UnitTests/Common/GVFSEnlistmentHealthTests.cs

```diff
@@ -0,0 +1,234 @@
+﻿using GVFS.Common;
+using GVFS.Tests.Should;
+using NUnit.Framework;
+using System.Collections.Generic;
+
+namespace GVFS.UnitTests.Common
+{
+    [TestFixture]
+    public class GVFSEnlistmentHealthTests
+    {
+        private readonly char sep = GVFSPlatform.GVFSPlatformConstants.PathSeparator;
+
+        private List<string> fauxGitFiles = new List<string>();
+        private List<string> fauxGitFolders = new List<string>();
+        private List<string> fauxPlaceholderFiles = new List<string>();
+        private List<string> fauxPlaceholderFolders = new List<string>();
+        private List<string> fauxModifiedPathsFiles = new List<string>();
+        private List<string> fauxModifiedPathsFolders = new List<string>();
+        private List<string> fauxSkipWorktreeFiles = new List<string>();
+        private GVFSEnlistmentHealthCalculator enlistmentHealthCalculator;
+        private GVFSEnlistmentHealthCalculator.GVFSEnlistmentHealthData enlistmentHealthData;
+
+        [TestCase]
+        public void OnlyOneHydratedDirectory()
+        {
+            this.fauxGitFiles.AddRange(new string[] { "A/1.js", "A/2.js", "A/3.js", "A/4.js", "B/1.js", "B/2.js", "B/3.js", "B/4.js", "C/1.js", "C/2.js", "C/3.js", "C/4.js" });
+            this.fauxPlaceholderFiles.AddRange(new string[] { "A/1.js", "A/2.js", "A/3.js" });
+
+            this.enlistmentHealthData = this.GenerateStatistics(string.Empty);
+
+            this.enlistmentHealthData.DirectoryHydrationLevels[0].Key.ShouldEqual("A");
+            this.enlistmentHealthData.DirectoryHydrationLevels[0].Value.ShouldEqual(0.75m);
+            this.enlistmentHealthData.DirectoryHydrationLevels[1].Value.ShouldEqual(0);
+            this.enlistmentHealthData.DirectoryHydrationLevels[2].Value.ShouldEqual(0);
+            this.enlistmentHealthData.GitTrackedItemsCount.ShouldEqual(12);
+            this.enlistmentHealthData.PlaceholderCount.ShouldEqual(3);
+            this.enlistmentHealthData.ModifiedPathsCount.ShouldEqual(0);
+            this.enlistmentHealthData.PlaceholderPercentage.ShouldEqual(0.25m);
+            this.enlistmentHealthData.ModifiedPathsPercentage.ShouldEqual(0);
+        }
+
+        [TestCase]
+        public void AllEmptyLists()
+        {
+            this.enlistmentHealthData = this.GenerateStatistics(string.Empty);
+
+            this.enlistmentHealthData.GitTrackedItemsCount.ShouldEqual(0);
+            this.enlistmentHealthData.PlaceholderCount.ShouldEqual(0);
+            this.enlistmentHealthData.ModifiedPathsCount.ShouldEqual(0);
+            this.enlistmentHealthData.PlaceholderPercentage.ShouldEqual(0);
+            this.enlistmentHealthData.ModifiedPathsPercentage.ShouldEqual(0);
+        }
+
+        [TestCase]
+        public void OverHydrated()
+        {
+            this.fauxGitFiles.AddRange(new string[] { "A/1.js", "A/2.js", "A/3.js", "A/4.js", "B/1.js", "B/2.js", "B/3.js", "B/4.js", "C/1.js", "C/2.js", "C/3.js", "C/4.js" });

```

Comment:
I wonder if we can have a class for testing that has the git files setup and we call various method like `AddAllAsPlaceholders` etc. so it is easier to see what is being setup in the test.  I have really like using a [builder pattern](https://ardalis.com/improve-tests-with-the-builder-pattern-for-test-data) because it makes that tests much more readable for what is being tested.

Reply:
Tracking this in a larger GVFSEnlistmentHealhTests refactoring issue: https://github.com/microsoft/VFSForGit/issues/1299